lib/uplinkc/uplink_definitions.h

```diff
@@ -70,7 +91,48 @@ typedef struct EncryptionAccess {
     char key[32];
 } EncryptionAccess;
 
+typedef struct ObjectInfo {
+    uint32_t   version;
+    BucketInfo bucket;
+    char       *path;
+    bool       is_prefix;
+    char       *content_type;
+    int64_t    created;
+    int64_t    modified;
+    int64_t    expires;
+} ObjectInfo;
+
+typedef struct ObjectList {
+    char       *bucket;
+    char       *prefix;
+    bool       more;
+    ObjectInfo *items;
+    int32_t    length;
+} ObjectList;
+
 typedef struct UploadOptions {
     char    *content_type;
     int64_t expires;
 } UploadOptions;
+
+typedef struct ListOptions {
+    char    *prefix;
+    char    *cursor;
+    char    delimiter;
+    bool    recursive;
+    int8_t  direction;
+    int64_t limit;
+} ListOptions;
+
+typedef struct ObjectMeta {
```

Comment:
Wondering whether we could get rid of `ObjectMeta` altogether and just use `ObjectInfo`?

Reply:
i don't know if this is related, but i tend to prefer APIs where it's very clear what I can and can't set. I would much prefer keeping return types (that have more computed information) and argument types (strictly the things i can change) separated

Reply:
Yeah, I agree... but I'm trying to figure out the difference.

Reply:
I would expect the types be more like:

```
typedef struct ObjectList {
    char       *bucket;
    char       *prefix;
    bool       more;
    ObjectListItem *items;
    int32_t    length;
} ObjectList;

typedef struct ObjectListItem {
    bool       is_prefix;

    uint32_t   version;
    char       *path;
    char       *content_type;

    int64_t    created;
    int64_t    modified;
    int64_t    expires;
} ObjectListItem;

typedef struct ObjectInfo {
    uint32_t   version;
    char       *path;
    char       *content_type;

    int64_t    created;
    int64_t    modified;
    int64_t    expires;
    uint64_t   size;

    uint8_t    *checksum_bytes;
    uint64_t   checksum_length;
} ObjectInfo;
```

Reply:
Either way it seems that something can be improved in `lib/uplink` as well.

Reply:
I don't disagree with you guys on this but we were just trying to mirror the go api and this is what the go api looks like. Can we create a separate issue to change libuplink and address this design change?

Reply:
resolution: looks like there are quite a few bits that need improvements on `lib/uplink` side as well